src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -298,7 +314,99 @@ public static Type GetUnderlyingType(Type enumType)
         }
 
         public static TEnum[] GetValues<TEnum>() where TEnum : struct, Enum
-            => (TEnum[])GetValues(typeof(TEnum));
+        {
+            RuntimeType type = (RuntimeType)typeof(TEnum);
```

Comment:
This is slow API by design. It allocates a new array on every invocation. Is it really worth it paying trading generic code size expansion to make it faster?

Reply:
This changed it from being a relatively slow API to a relatively fast API that allocates 1/4th of what it did before. But yes, if anyone were using it on a hot path, we'd recommend they not because of that remaining array allocation, suggesting they do their own caching instead.

My original plan for this PR was just to fix Enum.IsDefined. GetName and GetNames came along for the ride as they were one-line fixes. And then I happened to benchmark GetValues and saw how much waste there was.  If you don't think GetValues is worth it, though, I can revert it.

As a middle-ground, we could also do just:
```C#
public static TEnum[] GetValues<TEnum>() where TEnum : struct, Enum
{
    RuntimeType type = (RuntimeType)typeof(TEnum);

    ulong[] ulValues = InternalGetValues(type);
    var results = new TEnum[ulValues.Length];

    for (int i = 0; i < ulValues.Length; i++)
    {
        results[i] = (TEnum)ToObject(typeof(TEnum), ulValues[i]);
    }

    return results;
}
```
It doesn't eliminate any of the boxing and still pays the cost of the ToObject, but it does avoid going through a non-generic array and most of the additional code size...

|    Method |           Toolchain |      Mean | Ratio | Allocated |
|---------- |-------------------- |----------:|------:|----------:|
| GetValues | \master\corerun.exe | 573.06 ns |  1.00 |     224 B |
| GetValues |     \pr\corerun.exe |  19.05 ns |  0.03 |      56 B |
| GetValues |     \pr2\corerun.exe | 380.31 ns |  0.67 |      224 B |

?

Reply:
My estimate is that the generic expansion will cost about 0.5ms extra in JIT and type loading (assuming non-AOT case). It means that one has to call the API about 1000 times for given `T` to break even. If somebody is calling this API 1000 times for given `T`, they either do not care about performance at all or they should be caching the result.

Reply:
> My estimate is that the generic expansion will cost about 0.5ms extra in JIT and type loading (assuming non-AOT case)

How did you reach that estimate?  This method is already generic and would be specialized; the difference between:
```C#
(TEnum[])GetValues(typeof(TEnum))
```
and
```C#
public static TEnum[] GetValues<TEnum>() where TEnum : struct, Enum
{
    RuntimeType type = (RuntimeType)typeof(TEnum);

    ulong[] ulValues = InternalGetValues(type);
    var results = new TEnum[ulValues.Length];

    for (int i = 0; i < ulValues.Length; i++)
    {
        results[i] = (TEnum)ToObject(typeof(TEnum), ulValues[i]);
    }

    return results;
}
```
costs ~500us of JIT time? I agree that's not worth it, but it's also way larger than I'd have expected.

Reply:
My estimate was for the original larger method.